icu4c/source/common/localematcher.cpp

```diff
@@ -605,10 +607,11 @@ class LocaleLsrIterator {
 
 const Locale *LocaleMatcher::getBestMatch(const Locale &desiredLocale, UErrorCode &errorCode) const {
     if (U_FAILURE(errorCode)) { return nullptr; }
-    int32_t suppIndex = getBestSuppIndex(
+    std::optional<int32_t> suppIndex = getBestSuppIndex(
         getMaximalLsrOrUnd(likelySubtags, desiredLocale, errorCode),
         nullptr, errorCode);
-    return U_SUCCESS(errorCode) && suppIndex >= 0 ? supportedLocales[suppIndex] : defaultLocale;
+    return U_SUCCESS(errorCode) && suppIndex.has_value() ? supportedLocales[suppIndex.value()]
+                                                         : defaultLocale;
```

Comment:
optional -- and this pun seems unavoidable:

std::optional has an `operator bool` which does the same thing as `has_value()`.
https://en.cppreference.com/w/cpp/utility/optional/operator_bool

I don't know why they created a new class with two ways of doing the same thing, but I suspect that using the optional value in a boolean expression is intended to be colloquial. Although it might be less obvious/readable, so I won't insist on it. But it would make the code a little shorter and spiffier.

Example:
```suggestion
    return U_SUCCESS(errorCode) && suppIndex ? supportedLocales[suppIndex.value()] : defaultLocale;
```